django/http/request.py

```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
             location = '//%s' % self.get_full_path()
         bits = urlsplit(location)
         if not (bits.scheme and bits.netloc):
-            current_uri = '{scheme}://{host}{path}'.format(scheme=self.scheme,
-                                                           host=self.get_host(),
-                                                           path=self.path)
-            # Join the constructed URL with the provided location, which will
-            # allow the provided ``location`` to apply query strings to the
-            # base path as well as override the host, if it begins with //
-            location = urljoin(current_uri, location)
+            # Handle the most common and trivial case trivially. If the location is
+            # absolute and no scheme or host (netloc) is provided we can shortcut
+            # urljoin as long as no path segments are '.' or '..'.
+            if (bits.path.startswith('/') and not bits.scheme and not bits.netloc and
+                    '/./' not in bits.path and '/../' not in bits.path):
```

Comment:
This check doesn't catch a location that ends with `/.` or `/..`

Reply:
Correct, I might need to revert that bit.

Reply:
I knew it'd miss that when I threw it out there, but I assume the slow path (going through `urljoin`) handles that. 
IMHO it's yet another edge case, and a harder one to catch (I think the quickest way to do it would be `bits.path[-3:] == "/.."` and `bits.path[-2:] == '/.'` maybe?)